### ISSUE TITLE
Strapi upload file name.

### ISSUE THREAD

**Issue by rinkusam12** [strapi/strapi#3622]
Hello,
Is there a way to change the default behaviour for upload file name, i.e. currently the file get stored with the hash id as its file name, but I want to have it stored with its original file name.
for example: 
the file name with example.jpg should be stored as https://test.s3.amazonaws.com/example.jpg,   and not as https://test.s3.amazonaws.com/113a96e0c7a0475ea85aa919c7097717.jpg

**Comment by derrickmehaffy** [comment]
@rinkusam12 please use the strapi slack ( https://slack.strapi.io ) or stack overflow for questions. We try to keep the github issues clean by only allowing for bug reports, feature requests, and technical design discussions.

For now I will mark this issue as closed, however to answer your question, there is no current way to configure this without code modification. The reason for the hash id is to prevent possible duplicate names on different files, this is quite common in storage APIs for uploaded files.

**Comment by lauriejim** [comment]
You can deal with that by updating the `buffurize` function of the **Upload**  plugin.

Please for that you will have to check how the **extensions** folder work.
**📚Documentation here:** https://strapi.io/documentation/3.0.0-beta.x/concepts/concepts.html#extensions

And after you will have to find the path of the file you want to update.
It will be this one https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/services/Upload.js
So you will have to create an empty file at the same path in the extension folder.
It will be `./extensions/upload/services/Upload.js`
```js
module.exports = {
};
```

In this file you will have to create the function you want to update/override.
It will be this function https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/services/Upload.js#L26
So you have to copy this function and past it in your extensions file.

When it's done you can modify the function as you want.

And in your case you will have to modify this key value https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/services/Upload.js#L48
Replace the `uuid` stuff by the `stream.name`

**Comment by emal-nawabi** [comment]
Hi,
I also want to upload plugin that preserves the file name. I go through your guidance but it does not work. Could you please @lauriejim provide the full modification code that is needed to just change the behaviour, I mean to preserve the file name. As I wouldn't be able to modify that part of the code properly.

**Comment by laurenskling** [comment]
I'd like to be able to change this behaviour as well. I don't need hashing on my images, Gatsby will do that for me. But Gatsby won't allow me to change filenames, so now I'm stuck with hashed filenames which can't be optimized for SEO.

**Comment by derrickmehaffy** [comment]
@laurenskling this would need to be a manual modification you apply to your application, we evaluated this years ago and decided not to change this logic as it basically is the only method for handling duplication (think if some users took a picture or a screenshot on their phone. All phones basically follow the same naming system so it's entirely possible two users take two different pictures with the same name.)

Basically the hash key appended to the end is to ensure one picture wouldn't conflict with another, as it stands right now Strapi will not change this logic as we build features the masses and this a data loss prevention feature.

**Comment by laurenskling** [comment]
Using Cloudinary, cloudinary will safe guard me from have duplicate file names. 
So wouldn't it be better to have this handled in the provider? So the cloudinary provider could have an option to not do it on the Strapi side?

**Comment by derrickmehaffy** [comment]
> Using Cloudinary, cloudinary will safe guard me from have duplicate file names. 
> So wouldn't it be better to have this handled in the provider? So the cloudinary provider could have an option to not do it on the Strapi side?

This would complicate the logic and would also be a breaking change, this is something we could possibly look at for v5 but not right now.

**Comment by sebscholl** [comment]
In case people are still affected by this, I was able to figure it out.

The code is here that enables the feature: https://github.com/strapi/strapi/blob/0072ce9789eb1aeae8662ea7dd9478a8952530fc/packages/core/upload/server/services/upload.js#L35

In my `src/index.js` file I was able to override the exported function that uses this method, like so:

```ts
// src/index.js
module.exports = {
  register({ strapi }) {
    /**
     * Override the default upload info formatter with custom function.
     */
    strapi.services["plugin::upload.upload"].formatFileInfo =
      require("./extensions/upload/overrides").formatFileInfoOverride;
  },
}
``` 

I made a file to store the override function at `src/extensions/upload/overrides`, which keeps the `src/index.js` file cleaner.

This is the exported override function, which is really just the original function but slightly refactored.

```js
// src/extensions/upload/overrides
const _ = require("lodash");
const path = require("path");
const { extension } = require("mime-types");

/**
 * Overriding formatFileInfo function for upload plugin to
 * maintain original file naming.
 */
export async function formatFileInfoOverride(
  { filename, type, size },
  fileInfo = {},
  metas = {}
) {
  const fileService = strapi.plugin("upload").service("file");

  let ext = path.extname(filename);

  if (!ext) {
    ext = `.${extension(type)}`;
  }
  const usedName = (fileInfo.name || filename).normalize();
  const basename = path.basename(usedName, ext);

  const entity = {
    ext,
    mime: type,
    hash: basename,
    name: usedName,
    folder: fileInfo.folder,
    caption: fileInfo.caption,
    alternativeText: fileInfo.alternativeText,
    size: Math.round((size / 1000) * 100) / 100,
    folderPath: await fileService.getFolderPath(fileInfo.folder),
  };

  const { refId, ref, field } = metas;

  if (refId && ref && field) {
    entity.related = [
      {
        id: refId,
        __type: ref,
        __pivot: { field },
      },
    ];
  }

  if (metas.path) {
    entity.path = metas.path;
  }

  if (metas.tmpWorkingDirectory) {
    entity.tmpWorkingDirectory = metas.tmpWorkingDirectory;
  }

  return entity;
}
```

I've really been impressed with strapi, but it's been pretty difficult to work with when handling files. I feel like it was built with a single use case in mind, and their isn't very clean API's (or maybe just documentation) for extensibility.

Anyways, this works.

**Comment by uvarov-frontend** [comment]
If anyone finds it useful, I’m sharing my solution for deleting/changing the hash for files.

Create/change `src/extensions/upload/strapi-server.ts`:
```
export default (plugin) => {
  const replaceFileName = () => {
    return {
      ...uploadServices({ strapi }),
      formatFileInfo,
    };
  };

  plugin.services['upload'] = replaceFileName;
  return plugin;
};
```

Create a file and add `src/extensions/upload/formatFileInfo.ts`:
```
const { nameToSlug } = require('@strapi/utils');
const path = require("path");
const { extension } = require("mime-types");

const generateFileName = (name) => {
  const baseName = nameToSlug(name, { separator: '_', lowercase: false });

  # Change baseName
  return `${baseName};`
};

const formatFileInfo = async ({ filename, type, size }, fileInfo = {}, metas = {}) => {
	const fileInfoAny = fileInfo as any;
	const metasAny = metas as any;
	const fileService = strapi.plugin("upload").service("file");

	let ext = path.extname(filename);
	if (!ext) {
		ext = `.${extension(type)}`;
	}
	const usedName = (fileInfoAny.name || filename).normalize();
	const basename = path.basename(usedName, ext);

	const entity: any = {
		name: usedName,
		alternativeText: fileInfoAny.alternativeText,
		caption: fileInfoAny.caption,
		folder: fileInfoAny.folder,
		folderPath: await fileService.getFolderPath(fileInfoAny.folder),
		hash: generateFileName(basename),
		ext,
		mime: type,
		size: Math.round((size / 1000) * 100) / 100,
	};

	const { refId, ref, field } = metasAny;

	if (refId && ref && field) {
		entity.related = [
			{
				id: refId,
				__type: ref,
				__pivot: { field },
			},
		];
	}

	if (metasAny.path) {
		entity.path = metasAny.path;
	}

	if (metasAny.tmpWorkingDirectory) {
		entity.tmpWorkingDirectory = metasAny.tmpWorkingDirectory;
	}

	return entity;
}

export default formatFileInfo;

```

**Comment by raysuhyunlee** [comment]
In case of reading this in 2025, `generateFileName` has been moved to a separate file in [this PR](https://github.com/strapi/strapi/commit/7de63df3ae739fecd8473832676dce2b895eea52).

You no longer need to override `formatFileInfo`. Instead, you can simply override `generateFileName` like this:
```ts
// ./src/extensions/upload/strapi-server.ts

export default (plugin) => {
    plugin.services['image-manipulation'].generateFileName = (name: string) => {
        return `custom-file-name-${name}`
    };

    return plugin
}
```
